### PR TITLE
net: socket timeout options

### LINKED ISSUE #11820: sockets: Implement (POSIX-compatible) timeout support
It's not possible to implement a robust networking solution without timeouts. So, sockets should implement them, following the POSIX mandated API.

### PR DESCRIPTION
**Pull request by hubertmis**
This patch adds SO_RCVTIMEO and SO_SNDTIMEO options used to time out
socket receiving or sending operations.

This PR partially fixes #11820 (only receive part).

Signed-off-by: Hubert Miś <hubert.mis@nordicsemi.no>

### CODE REVIEW

**Comment by pfalcon** [comment]
Thanks! Please add `Fixes: #11820` (assuming complete enough support is implemented). It's also a common recommendation to have tests as a 2nd commit, to avoid large, hard to review patches.

**Comment by hubertmis** [comment]
> Please add `Fixes: #11820` (assuming complete enough support is implemented).

Done. I hope it's complete enough. I've created complete tests for recv functions, but could not find a way to time out send in tests.

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good, but we need to have these options to support other socket types too

**Comment by hubertmis** [comment]
> we need to have these options to support other socket types too

I've added this option to packet and CAN sockets.

It seems that with net_mgmt this option cannot be set and TLS sockets handle timeouts in a different way. I'm not sure if this option can be supported with these socket implementations.

**Review by jukkar — APPROVED** [review]
Thanks, LGTM

**Review by rlubos — APPROVED** [review]
LGTM

File: tests/net/socket/udp/src/main.c
Comment (pfalcon) [inline tests/net/socket/udp/src/main.c:799]: We never use EWOULDBLOCK in the stack, instead having standardized on EAGAIN. Can you please replace all occurrences to preserve consistency?

**Review by pfalcon — CHANGES_REQUESTED** [review]
Error codes consistency concern.

**Comment by pfalcon** [comment]
@hubertmis: How well did you test timeout error reporting. EAGAIN should be returned in this case. I see that your test checks for that (well, for EWOULDBLOCK, I requested change to EAGAIN for consistency). At the same time, we have code like https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/ip/net_context.c#L2049, which explicitly converts underlying EAGAIN to network ETIMEDOUT. Such a conversion is arguable a mistake in the first place, as there was proposal to stick with POSIX semantics, and IIRC, POSIX doesn't use ETIMEDOUT is socket subsys.

**Comment by pfalcon** [comment]
> IIRC, POSIX doesn't use ETIMEDOUT is socket subsys.

Ok, https://pubs.opengroup.org/onlinepubs/9699919799/ does talk about ETIMEDOUT, but it's described as a fatal error in the underliyng connection, e.g. for recv():

> [ETIMEDOUT]
    The connection timed out during connection establishment, or due to a transmission timeout on active connection.

OTOH, `man 7 socket` says:

> SO_RCVTIMEO and SO_SNDTIMEO
              Specify  the  receiving  or  sending  timeouts until reporting an error.  The argument is a struct timeval.  If an input or output function blocks for this period of time, and data has been sent or received, the
              return value of that function will be the amount of data transferred; if no data has been transferred and the timeout has been reached, then -1 is returned with errno set to EAGAIN or EWOULDBLOCK, or EINPROGRESS
              (for  connect(2)) just as if the socket was specified to be nonblocking.

@hubertmis: Bottom line, can you please re-check this stuff. And could you try to add recv() SO_RCVTIMEO test for TCP socket?

**Comment by pfalcon** [comment]
@hubertmis: Another thing, as you can see, the man page talks about SO_RCVTIMEO/SO_SNDTIMEO being applied to connect(). I understand it would be too much of scope creep to ask you to do something about that, but I think that we at least should document somehow that SO_RCVTIMEO/SO_SNDTIMEO apply only to recv()/send() (ahem, did you read https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/net_context.h#L910 ?), and not to e.g. connect(), like in Linux.

**Comment by hubertmis** [comment]
@pfalcon, Thanks for your detailed review and comments. I created this PR, because I need `SO_RCVTIMEO` support for UDP sockets. Implementation was simple enough for me and I shared my results. I don't know any details yet about Zephyr network internals and didn't check thoroughly what underlying functions are called.

> the man page talks about SO_RCVTIMEO/SO_SNDTIMEO being applied to connect()

I'll look around `connect()`, looks like a nice challenge for winter evenings. If I don't succeed, I'll add note to `SO_RCVTIMEO`.

> did you read https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/net_context.h#L910 ?

I'll also look at `net_context_send()`. If I can't implement timeout inside, I think it would be wise to remove `SO_SNDTIMEO` part from this PR and leave `SO_RCVTIMEO`.

> can you please re-check this stuff. And could you try to add recv() SO_RCVTIMEO test for TCP socket?

I've based my tests on socket options description that you can find here https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_16 . I understand that `EAGAIN` should be used if there is no data available to receive when timeout expires, while `ETIMEDOUT` should be reported if connection established failed because of timeout. I'll try to add test for `recv()` TCP.

**Comment by hubertmis** [comment]
> I'll look around `connect()`, looks like a nice challenge for winter evenings. If I don't succeed, I'll add note to `SO_RCVTIMEO`

Actually, the description of `SO_RCVTIMEO` states:

> If a receive operation has blocked for this much time without receiving additional data, it shall return with a partial count or errno shall be set to [EAGAIN] or [EWOULDBLOCK] if no data were received.

But it does not cover connection establishment procedure. I don't think this timeout value should affect `connect()` in any way

**Comment by hubertmis** [comment]
> If I can't implement timeout inside, I think it would be wise to remove `SO_SNDTIMEO` part from this PR and leave `SO_RCVTIMEO`.

I reviewed sending path. It looks that sending process can block in multiple places:
1. Packet allocation
2. TCP may be implemented asynchronously
3. Network interface sending procedure may be blocking

I can see two options to implement timeout for sending path:
1. Subtract part of timeout for each blocking procedure and pass decreasing timeout from one blocking procedure to another (if timeout is set to 2 seconds and packet allocation took 500 ms, pass 1.5 s timeout to TCP).
2. Start a timer when send operation starts and when the timer expires, terminate ongoing TX.

Unfortunately, I don't have enough knowledge or experience to select or implement any of such solutions. Because of that, I'll simply remove `SO_SNDTIMEO` option to avoid confusion of socket users.

**Comment by hubertmis** [comment]
On the other hand, it looks that packets in socket send path are added to a queue before they reach interface. Because of that blocking interface should not cause blocking `net_context_send()`.

Also I've skimmed TCP implementations and it looks that send path is not blocking here as well.

Perhaps using timeout in `pkt = context_alloc_pkt(context, len, PKT_WAIT_TIME);` in `context_sendto()` function would be enough. @pfalcon, what do you think?

**Comment by pfalcon** [comment]
> I'll look around connect(), looks like a nice challenge for winter evenings. If I don't succeed, I'll add note to SO_RCVTIMEO.

To clarify, as I mentioned, I don't look to scope-creep your patch. It would be nice to get it into the mainline before 2.5 freeze (which means this week). But then I hope you agree that previously, there was no SO_RCVTIMEO/SO_SNDTIMEO defined at all, so someone using them, would get compile-time error, and knew that "nothing works". With you patch, it may be subtle situation that someone ports code from another system (like Linux), where SO_RCVTIMEO applies to connect(), but in Zephyr, it (silently) doesn't. Debugging such cases may be hellish. So again, I at least propose to add a quick note, e.g. here: https://github.com/zephyrproject-rtos/zephyr/pull/31369/files#diff-36069eb2479b89a497701aae06f843dd18c0bd3e182a052efac86c228b2bde00R799, something like "sockopt: Receive timeout. Applies only to recv(), etc., and not to connect()".

> I'll also look at net_context_send(). If I can't implement timeout inside, I think it would be wise to remove SO_SNDTIMEO part from this PR and leave SO_RCVTIMEO.

I'd say, that's not the desired outcome. Your implementation of SO_SNDTIMEO seems to be correct. It's underlying network stack which doesn't support it (according to docs, which might be stale). I'd say, leave it in place, but maybe add to commit message a note about that (but first check net_context_send(), maybe it's not fully actual).

**Comment by pfalcon** [comment]
> I've based my tests on socket options description that you can find here https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_10_16 . I understand that EAGAIN should be used if there is no data available to receive when timeout expires, while ETIMEDOUT should be reported if connection established failed because of timeout. I'll try to add test for recv() TCP.

You see, that's pretty formal description there. "recv()" and "connection establishment" also don't exactly mix in a clear way, without further clarification paragraphs. Because you see, connect() deals with "connection establishment", and normally when it returns, it's established. So, talking about "connection establishment" with recv() a bit strange.

Again, my understanding:
* EAGAIN is returned on transient condition, like "not data currently available/can be sent".
* ETIMEDOUT is returned when fatal condition is detected with the underlying connection. Socket is no longer active, not further data operation will succeed. In case of TCP, that can be e.g.: there was a previous send() operation, and there was ultimate timeout in getting ACK for that data. System decided that connection is irreparably broken, and marked it as such. When this decision occurred with recv() operation pending, the system will terminate it with ETIMEDOUT condition. Note that this "ultimate timeout" can be hours, years, of forever.

**Comment by pfalcon** [comment]
> On the other hand, it looks that packets in socket send path are added to a queue before they reach interface. Because of that blocking interface should not cause blocking net_context_send().

Well, it should be clear when blocking would happen - if "queue is full" (or more realistically, when you can't get a free buffer to put data into to add to the queue). Blocking socket under such conditions should wait for free buffer. And under some circumstances, it may end up waiting forever. That's exactly what we'd like to avoid with send timeout. So, *any* option to implement send timeout is better than not implementing it at all.

> Perhaps using timeout in pkt = context_alloc_pkt(context, len, PKT_WAIT_TIME); in context_sendto() function would be enough. @pfalcon, what do you think?

Per the above, I'd consider that to be a noticeable improvement over the current situation. (Of course, if it works as expected ;-) ).


Thanks for looking into this!

**Comment by hubertmis** [comment]
@pfalcon, I've added tests for TCP `recv()`. I'll add a note to `SO_RCVTIMEO`.

I'm also going to work a little around send packet allocator timeout, I'll try to add some tests here as well. But I'm not sure if I'll have enough time this week.

I think I could split this PR: right now `SO_RCVTIMEO` part seems pretty good, and `SO_SNDTIMEO` part could be a part of separated PR, merged later, when proven to work.

**Comment by jukkar** [comment]
> I can see two options to implement timeout for sending path:
> 
>     1. Subtract part of timeout for each blocking procedure and pass decreasing timeout from one blocking procedure to another (if timeout is set to 2 seconds and packet allocation took 500 ms, pass 1.5 s timeout to TCP).
> 
>     2. Start a timer when send operation starts and when the timer expires, terminate ongoing TX.

The 1. is quite complicated and IMHO not really needed. Currently the sending timeout is not used in net_context itself, we pass the value to offloading driver but the native code ignores it. There was some idea to use it in `net_context_send()` but it was abandoned after the API was created. I would question the use of the sending timeout, what is the purpose of the timeout in sending side. For UDP packets, the packet might be lost in transit anyway, and for TCP data there would be TCP timeouts that handle it.
On the other hand, the `SO_SNDTIMEO` could be very useful in `connect()`.

**Comment by hubertmis** [comment]
> what is the purpose of the timeout in sending side

I believe there are cases where application would like to have control how long `send()` call is blocked until it decides to drop the frame (when buffers are unavailable, or network driver is stuck and does not accept frames to its queue). Currently, send function uses constant 1 second timeout when buffer are unavailable.

However, I don't need send timeout for my applications. Because of that I removed `SO_SNDTIMEO` from this PR. After we merge this, I'll open a new one with `SO_SNDTIMEO` implementation, where we can discuss if it is needed.

**Comment by jukkar** [comment]
> Because of that I removed `SO_SNDTIMEO` from this PR. After we merge this, I'll open a new one with `SO_SNDTIMEO` implementation, where we can discuss if it is needed.

Thanks, that is a good plan :+1:

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
@hubertmis: Quick note while I'm re-reviewing the latest version: would you mind to change generic "net: socket timeout options" (still plural) to a more concrete: "net: socket: Implement SO_RCVTIMEO timeout option". If that's not much trouble of course.

**Comment by hubertmis** [comment]
> would you mind to change generic "net: socket timeout options" (still plural) to a more concrete: "net: socket: Implement SO_RCVTIMEO timeout option"

Thanks for catching it! I've missed changing first lines of commit messages while rebasing. I'll update in the evening.

File: subsys/net/ip/Kconfig
Comment (pfalcon) [inline subsys/net/ip/Kconfig:558]: I'd say "configured time" is ambiguous here: we're in Kconfig, so user seeing that will likely get question "where can I configure that time here?".

Something like: "Support setsockopt(SO_RCVTIMEO) call to set timeout for socket receive operations".

But formulated like that, makes us wonder if this option belongs to net/ip/Kconfig, or maybe net/lib/sockets/Kconfig (note that not I started it - you used the word "socket" yourself). The option still may be named generically, like "NET_OPT_RCVTIMEO".

Again, that's just a humble suggestion from a thorough review. Act only if you agree that would be an improvement. And I'd say, wait a possible nak from @jukkar, as he may have different considerations. I gave mine: if this option talks about "time out socket receive operation", then a user would expect to find it in socket configuration.

**Comment by pfalcon** [comment]
@hubertmis:

> I believe there are cases where application would like to have control how long send() call is blocked until it decides to drop the frame (when buffers are unavailable, or network driver is stuck and does not accept frames to its queue). Currently, send function uses constant 1 second timeout when buffer are unavailable.

This is not correct implementation in the first place. Again, non-blocking socket should wait for data to be sent for "as long as needed". OTOH, **the system** should perform fair resource planning and scheduling, to preclude infinite waits due to *deadlocks in resource (like buffers) access*. (We are pretty far from that picture in Zephyr so far).

 > However, I don't need send timeout for my applications. Because of that I removed SO_SNDTIMEO from this PR. After we merge this, I'll open a new one with SO_SNDTIMEO implementation, where we can discuss if it is needed.

As I mentioned, that's not the effect I wanted to achieve. But it's effectively the approach I follow (after multi-year experience in Zephyr). For example, #11820 is open for quite some time (and was known before it was ticketed), but I never had a good case to go and implemented, so it just sat there, "gathering momentum". So, can't complain you doing the same ;-).

> I'll update in the evening.

Ok, thanks for all the changes, and the work overall! There're a couple of cosmetic suggestions from me for your consideration. I'll keep my -1 for now, to avoid premature merge before you had a chance to look into them. But will +1 before tomorrow too.

File: subsys/net/ip/Kconfig
Comment (hubertmis) [reply subsys/net/ip/Kconfig:558]: I agree that "configured time" is ambiguous. However I think that this option should remain in IP Kconfig, alongside other options utilized by net context module. I'll rephrase help text to make it socket-less.

**Review by pfalcon — APPROVED** [review]

**Pull request merged**
